File: 2024/go/day1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

var (
	leftList  []int
	rightList []int
	diffList  []int
	simList   []int
)

func main() {
	readFile, err := os.Open("./input.txt")

	if err != nil {
		fmt.Println(err)
	}

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	for fileScanner.Scan() {
		line := strings.Split(fileScanner.Text(), "   ")
		leftList = append(leftList, convert(line[0]))
		rightList = append(rightList, convert(line[1]))
	}

	readFile.Close()

	sort.Ints(leftList)
	sort.Ints(rightList)

	for i := 0; i < len(leftList); i++ {
		var distance int
		if leftList[i] < rightList[i] {
			distance = rightList[i] - leftList[i]
		} else {
			distance = leftList[i] - rightList[i]
		}
		diffList = append(diffList, distance)
	}

	totalDist := 0

	for _, num := range diffList {
		totalDist = totalDist + num
	}

	fmt.Printf("Total Distance: %d\n", totalDist)

	// Part Two

	count := 0

	for _, num := range leftList {
		for _, num2 := range rightList {
			if num == num2 {
				count++
			}
		}
		simNum := num * count
		simList = append(simList, simNum)
		count = 0
	}

	totalSim := 0
	for _, num := range simList {
		totalSim = totalSim + num
	}

	fmt.Printf("Total Similarity Value: %d", totalSim)
}

func convert(value string) int {
	num, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		panic(err)
	}

	return num
}

File: 2024/ruby/day1/main.rb
def main
  file = File.open("./input.txt")
  data = file.readlines.map(&:chomp)
  file.close

  left_list = []
  right_list = []
  diff_list = []
  sim_list = []

  data.each do |line|
    d = line.split
    left_list.append(d[0].to_i)
    right_list.append(d[1].to_i)
  end

  left_list = left_list.sort
  right_list = right_list.sort

  left_list.each_with_index do |n, index|
    distance = 0
    if left_list[index] < right_list[index]
      distance = right_list[index] - left_list[index]
    else
      distance = left_list[index] - right_list[index]
    end
    diff_list.append(distance)
  end

  total_dist = 0

  diff_list.each do |num|
    total_dist = total_dist + num
  end

  puts("Total Distance: #{total_dist}\n")

  # Part Two

  count = 0

  left_list.each do |n|
    right_list.each do |i|
      if n == i
        count = count + 1
      end
    end
    sim_num = n * count
    sim_list.append(sim_num)
    count = 0
  end

  total_sim = 0
  sim_list.each do |num|
    total_sim = total_sim + num
  end

  puts("Total Similarity Value: #{total_sim}\n")
end

main
